File: data/js/analyzeProperties.js
require("dotenv").config();
const mongoose = require("mongoose");

// Create a simple schema for properties (you can import your existing model if available)
const propertySchema = new mongoose.Schema({}, { strict: false });
const Property = mongoose.model("Property", propertySchema, "properties");

const MONGO_URI = process.env.MONGO_URI;
if (!MONGO_URI) {
  console.error("MONGO_URI not defined in .env");
  process.exit(1);
}

mongoose
  .connect(MONGO_URI)
  .then(() => {
    console.log("Connected to MongoDB");
    runAnalytics();
  })
  .catch((err) => {
    console.error("Error connecting to MongoDB", err);
    process.exit(1);
  });

async function runAnalytics() {
  try {
    // Average price per city
    const avgPricePerCity = await Property.aggregate([
      {
        $group: {
          _id: "$city",
          avgPrice: { $avg: "$price" },
          count: { $sum: 1 },
        },
      },
      { $sort: { avgPrice: -1 } },
    ]);
    console.log("Average Price per City:");
    console.table(avgPricePerCity);

    // Distribution of properties by number of bedrooms
    const bedroomsDistribution = await Property.aggregate([
      { $group: { _id: "$bedrooms", count: { $sum: 1 } } },
      { $sort: { _id: 1 } },
    ]);
    console.log("Distribution by Bedrooms:");
    console.table(bedroomsDistribution);
  } catch (error) {
    console.error("Error during analytics", error);
  } finally {
    mongoose.connection.close();
  }
}

File: data/python/analyze_properties.py
import os
import logging
from pymongo import MongoClient
from dotenv import load_dotenv

# Load environment variables.
load_dotenv()

MONGO_URI = os.getenv("MONGO_URI")
if not MONGO_URI:
    logging.error("MONGO_URI is not set in .env")
    exit(1)

client = MongoClient(MONGO_URI)
db = client.get_default_database()
properties_collection = db["properties"]


def analyze_properties():
    """
    Aggregates property data by city and prints average, min, and max prices along with document counts.
    """
    pipeline = [
        {
            "$group": {
                "_id": "$city",
                "averagePrice": {"$avg": "$price"},
                "minPrice": {"$min": "$price"},
                "maxPrice": {"$max": "$price"},
                "count": {"$sum": 1}
            }
        },
        { "$sort": { "averagePrice": -1 } }
    ]

    results = properties_collection.aggregate(pipeline)
    print("Property Analysis by City:")
    for result in results:
        city = result["_id"] or "Unknown"
        avg_price = result["averagePrice"]
        min_price = result["minPrice"]
        max_price = result["maxPrice"]
        count = result["count"]
        print(f"City: {city}, Count: {count}, Avg Price: ${avg_price:.2f}, Min Price: ${min_price}, Max Price: ${max_price}")


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    analyze_properties()
    client.close()
